org.eclipse.dsp4j.jsonrpc/src/main/java/org/eclipse/dsp4j/jsonrpc/json/DebugMessageJsonHandler.java

```diff
@@ -0,0 +1,38 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Kichwa Coders Ltd. and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.dsp4j.jsonrpc.json;
+
+import java.util.Map;
+import java.util.function.Consumer;
+
+import org.eclipse.dsp4j.jsonrpc.adapters.DebugMessageTypeAdapterFactory;
+import org.eclipse.lsp4j.jsonrpc.json.JsonRpcMethod;
+import org.eclipse.lsp4j.jsonrpc.json.MessageJsonHandler;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.CollectionTypeAdapterFactory;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EitherTypeAdapterFactory;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EnumTypeAdapterFactory;
+
+import com.google.gson.GsonBuilder;
+
+public class DebugMessageJsonHandler extends MessageJsonHandler {
+	public DebugMessageJsonHandler(Map<String, JsonRpcMethod> supportedMethods) {
+		super(supportedMethods);
+	}
+
+	public DebugMessageJsonHandler(Map<String, JsonRpcMethod> supportedMethods, Consumer<GsonBuilder> configureGson) {
+		super(supportedMethods, configureGson);
+	}
+
+	public GsonBuilder getDefaultGsonBuilder() {

```

Comment:
This additional class could be avoided by passing a `Consumer<GsonBuilder>` to the default `MessageJsonHandler`. When multiple type adapter factories are configured for the same type, the last one should win.

Reply:
OK, the advantage of your approach too is that the other type factories don't need to be duplicated. A lot of the tests instantiate `DebugMessageJsonHandler` directly, so I'll probably want a method somewhere that constructs a `MessageJsonHandler` correctly for the tests in the same way as it is constructed for by the `DebugLauncher`. Perhaps a method in the `DebugLauncher` itself.

Reply:
I am having trouble writing a sensible method in DebugLauncher, at the moment it looks like the sample below, which seems uglier than having the extra class, ideas welcome.

The reason for the complication is the circular dependency between the `MessageJsonHandler` and the `DebugMessageTypeAdapterFactory`.

My next update will have the `DebugMessageJsonHandler` still, but rather than repeating everything in `getDefaultGsonBuilder` I'll call super's one and then add on the `DebugMessageTypeAdapterFactory`.

If you have a thought on something I overlooked, I would be pleased for the feedback.

```java

	static MessageJsonHandler createMessageHandler(Map<String, JsonRpcMethod> supportedMethods,
			Consumer<GsonBuilder> configureGson) {
		DebugMessageTypeAdapterFactory adapterFactory[] = new DebugMessageTypeAdapterFactory[] {null};
		Consumer<GsonBuilder> consumer = (builder) -> {
			builder.registerTypeAdapterFactory(adapterFactory[0]);
			configureGson.accept(builder);
		};
		MessageJsonHandler messageJsonHandler = new MessageJsonHandler(supportedMethods, consumer);
		adapterFactory[0] = new DebugMessageTypeAdapterFactory(messageJsonHandler);
		return messageJsonHandler;
	}
```

Reply:
Hmmm, we could change the `configureGson` parameter in MessageJsonHandler to `BiConsumer<GsonBuilder, MessageJsonHandler>`, but that's a breaking change. @akosyakov, @svenefftinge do you think this is worth it? Would help for cases when someone wants to customize parsing / serialization of messages.

Reply:
We could also add an additional constructor and deprecate the existing one. That would prevent it being a breaking change.

Reply:
Right, we don't even need to deprecate the old one.

Reply:
I have not made this change yet. The current version, and the new constructor we discussed suffers from a problem. It passes a reference to `this` from the constructor when the object is not fully created, so creating new API that required and actually enforced passing around an incompletely constructed object seems like a bad idea. If you would like me to pursue this further (e.g. add an `initialize()` method for example) I can raise and issue and work on that separately?